docs/extending/clusterprovider.md

```diff
@@ -0,0 +1,296 @@
+between steps.
+
+### Wizard pages and async
+
+Often you will need to call an asynchronous API (e.g. cloud API, invoking a tool through the
+shell) as part of composing a page.  To support this, the `showPage` method accepts a
+`Thenable<string>` such as a promise; the resulting HTML will be shown when the promise resolves.
+You can combine this with passing plain strings to display a 'wait' message. Here is an example:
+
+```javascript
+import { ClusterProviderV1 as cp } from 'vscode-kubernetes-tools-api';
+
+function next(wizard: cp.Wizard, action: cp.ClusterProviderAction, message: any): void {
+    if (message[cp.SENDING_STEP_KEY] === cp.SELECT_CLUSTER_TYPE_STEP_ID) {
+        wizard.showPage(`<h1>Fetching regions...</h1>`);
+        wizard.showPage(buildRegionPage());
+    } else {
+        // ...
+    }
+}
+
+async function buildRegionPage(): Promise<string> {
+    const regions = await contosocloud.listRegions();
+    const regionOptions = regions.map((r) => `<option value="${r.id}">${r.name}</option>`).join('\n');
+    const regionSelector = `<select name='region'>${regionOptions}</select>`;
+    const html = `<h1>Choose region</h1>
+        <form id='${cp.WIZARD_FORM_NAME}'>
+            <input type='hidden' name='${cp.SENDING_STEP_KEY}' value='${REGION_STEP_ID}' />
+            <input type='hidden' name='${cp.CLUSTER_TYPE_KEY}' value='${MY_PROVIDER_ID}' />
+            <p>Region: ${regionSelector}</p>
+            <button onclick='${cp.NEXT_PAGE}'>Next &gt;</button>
+        </form>
+        `;
+    return html;
+}
+```
+
+For pages that need to update multiple times in response to async operations (for example, a page
+that displays the progress of cluster creation), your page builder function can also return an
+`Observable<string>`.  **NOTE:** The `Observable` interface included in the Cluster Provider API
+is a custom interface, not the well known RxJS interface; please refer to the NPM package docs
+for details.
+
+## Registering the cluster provider
+
+In order to be displayed in the Add Existing Cluster or Create Cluster wizard, a cluster
+provider must be _registered_ with the Kubernetes extension.  This is the responsibility
+of the VS Code extension that hosts the cluster provider.  To do this, the extension must:
+
+* Activate in response to the Add Existing Cluster and/or Create Cluster commands
+* Request the Kubernetes extension's Cluster Provider API
+* Call the `register` method for each cluster provider it wants to display
+
+### Activating the cluster provider extension
+
+Your extension needs to activate in response to the Add Existing Cluster and/or Create
+Cluster commands, so that it can register its cluster provider(s) before the wizard is
+displayed.  To do this, your `package.json` must include the following activation events:
+
+```json
+    "activationEvents": [
+        "onCommand:extension.vsKubernetesConfigureFromCluster",
```

Comment:
interesting - i had no idea an extension could shadow activation events!

Reply:
Yeah, the activation event doesn't have to be from your own extension - I think VS Code just has global collections of commands, views, etc. and anything that knows the right name can take an interest in them.